Longread coverage workflow

weaglesBio opened the pull request:
Added long read coverage workflow and required modules for it.


<!--
# nf-core/treeval pull request

Many thanks for contributing to nf-core/treeval!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/treeval/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [X] This comment contains a description of changes (with reason).
- [] If you've fixed a bug or added code that should be tested, add tests!
  - [] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/treeval/tree/master/.github/CONTRIBUTING.md)
  - [] If necessary, also make a PR on the nf-core/treeval _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [X] Make sure your code lints (`nf-core lint`).
- [X] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [] Usage Documentation in `docs/usage.md` is updated.
- [] Output Documentation in `docs/output.md` is updated.
- [] `CHANGELOG.md` is updated.
- [] `README.md` is updated (including new tool citations and authors/contributors).

muffato commented on modules/local/longreadcoverage_graphoverall.nf:
```diff
@@ -0,0 +1,31 @@
+process LONGREADCOVERAGE_GRAPHOVERALL {
+    tag "$meta.id"
```
```suggestion
    tag "$meta.id"
    tag "process_single"
```

muffato commented on modules/local/sort_intersect.nf:
```diff
@@ -0,0 +1,23 @@
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    """
+    cat ${file} | sort ${args} > ${prefix}.bed
```
To make sure `sort` doesn't use more CPUs than allowed
```suggestion
    cat ${file} | sort --parallel=$task.cpus ${args} > ${prefix}.bed
```

muffato commented on modules/local/longreadcoverage_findhalfcoverage.nf:
I would remove "longreadcoverage" from the name of these three files. They're modules that exist on their own, and you may well use them in a different sub-workflow later.

muffato commented on modules/local/longreadcoverage_findhalfcoverage.nf:
```diff
@@ -0,0 +1,34 @@
+process LONGREADCOVERAGE_FINDHALFCOVERAGE {
+    tag "$meta.id"
```
```suggestion
    tag "$meta.id"
    label "process_single"
```

muffato commented on modules/local/get_reads_from_directory.nf:
```diff
@@ -0,0 +1,25 @@
+process GET_READS_FROM_DIRECTORY {
+    tag "${meta.id}"
+    label "process_low"
```
```suggestion
    label "process_single"
```

muffato commented on bin/findHalfcoverage.py:
```diff
@@ -0,0 +1,161 @@
+
+		if (coverage > arg5):
+
+			if (totalSize >250000):
```
Make this a command-line parameter (which may have 250000 as the default value) so that you can easily modify it from the pipeline

muffato commented on bin/findHalfcoverage.py:
```diff
@@ -0,0 +1,161 @@
+                  action="store", type="string", dest="depth", 
+                  help="depthgraph file, bp count at each depth")
+	parser.add_option("-w", "--wiggle", 
+                  action="store", type="string", dest="wig", 
```
The default value can be set right here, which makes it easier to maintain and keep the value in sync with the help message
```suggestion
                  action="store", type="string", dest="wig", default=5,
```

muffato commented on bin/findHalfcoverage.py:
```diff
@@ -0,0 +1,161 @@
+                  action="store", type="string", dest="wig", 
+                  help="wiggle room to add to depth cutoff ie 30X + wiggleroom.  Default is 5X")
+	parser.add_option("--cut", 
+                  action="store", type="string", dest="covcut", 
```
```suggestion
                  action="store", type="string", dest="covcut", default=60,
```

muffato commented on bin/findHalfcoverage.py:
```diff
@@ -0,0 +1,161 @@
+		print("Missing Options")
+		exit()
+
+	cov_filename        = options.covfile
+	mygenome_filename   = options.mygenome
+	depth_filename      = options.depth
+	wiggle_opt          = options.wig
+	cutoff_opt          = options.covcut
+
+	return (cov_filename, mygenome_filename, depth_filename, wiggle_opt, cutoff_opt)
+
+
+def main():
+# main program	
+
+	(arg1, arg2, arg3, arg4, arg5) = getArguments()
```
Pass `options` directly, so that you `main` will use `options.covfile` instead of `arg1`, `options.mygenome` instead of `arg2`, etc. That'll be much easier to maintain.

muffato commented on bin/findHalfcoverage.py:
```diff
@@ -0,0 +1,161 @@
+
+	(arg1, arg2, arg3, arg4, arg5) = getArguments()
+
+	if arg4 == None:
+		arg4 = 5
+
+	if arg5 == None:
+		arg5 = 60
+			
+
```
This would go away if you set the default values above
```suggestion
```

muffato reviewed (commented):
How can I test this ? All test profiles fail because of missing data

weaglesBio commented:
- Updated findHalfcoverage.py with parameter changes.
- Replaced SORT_INTERSECT with BEDTOOLS_SORT
- Replaced GET_FILES_FROM_DIRECTORY with GrabFiles process
- Added process single to GRAPHOVERALLCOVERAGE and FINDHALFCOVERAGE
- Removed LONGREADCOVERAGE prefix from local modules

We will be reviewing test data for automated testing shortly. Currently tested locally with both S and L size parameters.

muffato commented:
> We will be reviewing test data for automated testing shortly. Currently tested locally with both S and L size parameters.

What's the command-line for running the tests ?

muffato reviewed (commented):
Why did you replace `sort` with `bedtools sort` ? The authors themselves say it's not very useful, cf the disclaimer at the bottom of https://bedtools.readthedocs.io/en/latest/content/tools/sort.html
